vllm/core/scheduler.py

```diff
@@ -1583,6 +1583,43 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
 
         return self.scheduler_config.num_lookahead_slots
 
+    def _get_budget_for_chunk(self, budget: SchedulingBudget) -> int:
+        """When doing chunked prefill, calculate the token budget for a single
+        chunk. This dynamically scales the chunk size down as the number of
+        sequences that require prefilling increases. This ensures that a single
+        sequence with a very large prompt to prefill doesn't take the entire
+        remaining token budget, allowing other sequences to prefill and decode
+        concurrently."""
+
+        # First get the number of sequences that require prefill
+        prefilling_seqs = 0
+        for i in range(len(self.running)):
+            for seq in self.running[i].seqs:
+                if seq.data.stage == SequenceStage.PREFILL:
+                    prefilling_seqs += 1
+        # Queued requests will also need to prefill
+        prefilling_seqs += len(self.waiting)
```

Comment:
Since this function will be called by scheduler in every step, I'm afraid this loop would introduce latency overheads especially with high QPS.

Reply:
👍  It should be pretty simple to do something much faster like keep either a set or running counter around: `self.running_prefill_seqs`